File: src/automation-companions.ts
type JsonRecord = Record<string, unknown>;

interface BuildAutomationDescriptorInput {
  automation_id: string;
  title: string;
  owner: string;
  trigger_kind: string;
  target_surface_kind: string;
  summary: string;
  readiness_status: string;
  gate_policy: string;
  output_expectation: string[];
  target_command?: string | null;
  domain_projection?: JsonRecord | null;
}

interface BuildAutomationCatalogInput {
  summary: string;
  automations: ReturnType<typeof buildAutomationDescriptor>[];
  readiness_summary?: string | null;
}

function isRecord(value: unknown): value is JsonRecord {
  return typeof value === 'object' && value !== null && !Array.isArray(value);
}

function optionalString(value: unknown) {
  return typeof value === 'string' && value.trim() ? value.trim() : null;
}

function requireString(value: unknown, field: string) {
  const text = optionalString(value);
  if (!text) {
    throw new Error(`automation companion 缺少字符串字段: ${field}`);
  }
  return text;
}

function readStringList(value: unknown, field: string) {
  if (!Array.isArray(value)) {
    throw new Error(`automation companion 缺少数组字段: ${field}`);
  }
  return value.map((entry, index) => requireString(entry, `${field}[${index}]`));
}

export function buildAutomationDescriptor(input: BuildAutomationDescriptorInput) {
  return {
    surface_kind: 'automation_descriptor',
    automation_id: requireString(input.automation_id, 'automation_id'),
    title: requireString(input.title, 'title'),
    owner: requireString(input.owner, 'owner'),
    trigger_kind: requireString(input.trigger_kind, 'trigger_kind'),
    target_surface_kind: requireString(input.target_surface_kind, 'target_surface_kind'),
    summary: requireString(input.summary, 'summary'),
    readiness_status: requireString(input.readiness_status, 'readiness_status'),
    gate_policy: requireString(input.gate_policy, 'gate_policy'),
    output_expectation: readStringList(input.output_expectation, 'output_expectation'),
    ...(optionalString(input.target_command) ? { target_command: optionalString(input.target_command)! } : {}),
    ...(isRecord(input.domain_projection) ? { domain_projection: { ...input.domain_projection } } : {}),
  };
}

export function buildAutomationCatalog(input: BuildAutomationCatalogInput) {
  if (!Array.isArray(input.automations) || input.automations.length === 0) {
    throw new Error('automation catalog 需要至少一个 automation descriptor。');
  }
  return {
    surface_kind: 'automation',
    summary: requireString(input.summary, 'summary'),
    automations: input.automations.map((entry) => buildAutomationDescriptor(entry)),
    ...(optionalString(input.readiness_summary) ? { readiness_summary: optionalString(input.readiness_summary)! } : {}),
  };
}

File: python/opl_framework/automation_companions.py
from __future__ import annotations

from typing import Any, Mapping


def _non_empty_text(value: object) -> str | None:
    text = str(value or "").strip()
    return text or None


def _require_string(value: object, field: str) -> str:
    text = _non_empty_text(value)
    if text is None:
        raise ValueError(f"automation companion 缺少字符串字段: {field}")
    return text


def _require_string_list(value: object, field: str) -> list[str]:
    if not isinstance(value, list):
        raise ValueError(f"automation companion 缺少数组字段: {field}")
    return [_require_string(entry, f"{field}[{index}]") for index, entry in enumerate(value)]


def build_automation_descriptor(
    *,
    automation_id: str,
    title: str,
    owner: str,
    trigger_kind: str,
    target_surface_kind: str,
    summary: str,
    readiness_status: str,
    gate_policy: str,
    output_expectation: list[str],
    target_command: str | None = None,
    domain_projection: Mapping[str, Any] | None = None,
) -> dict[str, Any]:
    payload: dict[str, Any] = {
        "surface_kind": "automation_descriptor",
        "automation_id": _require_string(automation_id, "automation_id"),
        "title": _require_string(title, "title"),
        "owner": _require_string(owner, "owner"),
        "trigger_kind": _require_string(trigger_kind, "trigger_kind"),
        "target_surface_kind": _require_string(target_surface_kind, "target_surface_kind"),
        "summary": _require_string(summary, "summary"),
        "readiness_status": _require_string(readiness_status, "readiness_status"),
        "gate_policy": _require_string(gate_policy, "gate_policy"),
        "output_expectation": _require_string_list(output_expectation, "output_expectation"),
    }
    if _non_empty_text(target_command) is not None:
        payload["target_command"] = _non_empty_text(target_command)
    if isinstance(domain_projection, Mapping):
        payload["domain_projection"] = dict(domain_projection)
    return payload


def build_automation_catalog(
    *,
    summary: str,
    automations: list[Mapping[str, Any]],
    readiness_summary: str | None = None,
) -> dict[str, Any]:
    if not automations:
        raise ValueError("automation catalog 需要至少一个 automation descriptor。")
    payload: dict[str, Any] = {
        "surface_kind": "automation",
        "summary": _require_string(summary, "summary"),
        "automations": [dict(entry) for entry in automations if isinstance(entry, Mapping)],
    }
    if _non_empty_text(readiness_summary) is not None:
        payload["readiness_summary"] = _non_empty_text(readiness_summary)
    return payload
